nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java

```diff
@@ -45,52 +52,75 @@
-        compoundDoc.setPropertyValue("dc:title", compoundDocName);
-        compoundDoc = session.createDocument(compoundDoc);
-        createCompoundDocument(compoundDoc, archive);
-        return compoundDoc;
+        var parentDoc = session.getDocument(new PathRef(parent));
+        try (var zip = new ZipFile(archive.getFile())) {
+            List<ZipEntry> entries = zip.stream()
+                                        .filter(this::isAllowedEntry)
+                                        .sorted(Comparator.comparing(ZipEntry::getName))
+                                        .collect(Collectors.toList());
+            // Avoid giving the whole blobs
+            List<String> entryNames = entries.stream().map(ZipEntry::getName).collect(Collectors.toList());
+            String compoundType = runScript(COMPOUND_DOCTYPE_DETECTION_OPERATION, parentDoc,
+                    Map.of("entries", entryNames));
+            DocumentModel compoundDoc = session.createDocumentModel(parent, compoundDocName, compoundType);
+            compoundDoc.setPropertyValue("dc:title", compoundDocName);
+            compoundDoc = session.createDocument(compoundDoc);
+            createCompoundDocument(compoundDoc, zip, entries);
+            return compoundDoc;
+        } catch (IOException e) {
+            var message = String.format("Failed to create CompoundDocument for archive: %s in parent: %s",
+                    archive.getFilename(), parent);
+            throw new NuxeoException(message, e);
+        }
     }
 
-    protected void createCompoundDocument(DocumentModel compoundDoc, Blob archive) {
-        try (ZipFile zip = new ZipFile(archive.getFile())) {
-            zip.stream()
-               .filter(this::isAllowedEntry)
-               .sorted(Comparator.comparing(ZipEntry::getName))
-               .forEach(entry -> createEntry(compoundDoc, archive.getFilename(), zip, entry));
-        } catch (IOException e) {
-            String message = String.format(
-                    "Failed to create CompoundDocument: %s. something went wrong with archive: %s.",
-                    compoundDoc.getName(), archive.getFilename());
+    protected String runScript(String scriptId, DocumentModel doc, Map<String, ?> params) {
+        AutomationService automationService = Framework.getService(AutomationService.class);
+        try (OperationContext ctx = new OperationContext(doc.getCoreSession())) {
+            ctx.setInput(doc);
+            return (String) automationService.run(ctx, scriptId, params);
+        } catch (OperationException e) {
+            var message = String.format("Error while running script: %s for input: %s.", scriptId, doc);
             throw new NuxeoException(message, e);
         }
     }
 
+    protected void createCompoundDocument(DocumentModel compoundDoc, ZipFile zip, List<ZipEntry> entries)
+            throws IOException {
```

Comment:
As there're two/three lines in this method, it is worth merging it in the caller one.

Reply:
I had to make the doc final to use it in the forEach so I made compoundDoc then finalDoc